File: user_service/services/viacep_service.py
"""
Serviço de Integração ViaCEP
Sistema de busca automática de endereços por CEP
"""

import re
import aiohttp
import asyncio
from typing import Optional, Dict, Any
from dataclasses import dataclass


@dataclass
class AddressData:
    """
    Classe para representar dados de endereço do ViaCEP
    """
    cep: str
    logradouro: str
    complemento: str
    bairro: str
    localidade: str  # cidade
    uf: str  # estado
    ibge: str
    gia: str
    ddd: str
    siafi: str
    erro: bool = False


class ViaCEPService:
    """
    Serviço para integração com API ViaCEP
    
    Fornece métodos para busca de endereços por CEP
    com validação e tratamento de erros
    """
    
    BASE_URL = "https://viacep.com.br/ws"
    TIMEOUT = 10  # segundos
    
    @staticmethod
    def clean_cep(cep: str) -> str:
        """
        Remove formatação do CEP (hífens, espaços)
        
        Args:
            cep: CEP com ou sem formatação
            
        Returns:
            CEP apenas com números
        """
        if not cep:
            return ""
        
        # Remove tudo que não for dígito
        return re.sub(r'\D', '', str(cep))
    
    @staticmethod
    def format_cep(cep: str) -> str:
        """
        Formata CEP no padrão XXXXX-XXX
        
        Args:
            cep: CEP apenas com números
            
        Returns:
            CEP formatado ou string vazia se inválido
        """
        clean = ViaCEPService.clean_cep(cep)
        
        if len(clean) != 8:
            return ""
        
        return f"{clean[:5]}-{clean[5:]}"
    
    @staticmethod
    def validate_cep_format(cep: str) -> bool:
        """
        Valida formato do CEP (8 dígitos)
        
        Args:
            cep: CEP para validação
            
        Returns:
            True se formato válido, False caso contrário
        """
        clean_cep = ViaCEPService.clean_cep(cep)
        return len(clean_cep) == 8 and clean_cep.isdigit()
    
    @staticmethod
    async def fetch_address_by_cep(cep: str) -> Optional[AddressData]:
        """
        Busca endereço por CEP na API ViaCEP (assíncrono)
        
        Args:
            cep: CEP para busca
            
        Returns:
            AddressData se encontrado, None se erro
        """
        # Valida formato do CEP
        if not ViaCEPService.validate_cep_format(cep):
            return None
        
        clean_cep = ViaCEPService.clean_cep(cep)
        url = f"{ViaCEPService.BASE_URL}/{clean_cep}/json/"
        
        try:
            timeout = aiohttp.ClientTimeout(total=ViaCEPService.TIMEOUT)
            
            async with aiohttp.ClientSession(timeout=timeout) as session:
                async with session.get(url) as response:
                    if response.status == 200:
                        data = await response.json()
                        
                        # Verifica se o CEP foi encontrado
                        if 'erro' in data and data['erro']:
                            return None
                        
                        return AddressData(
                            cep=data.get('cep', ''),
                            logradouro=data.get('logradouro', ''),
                            complemento=data.get('complemento', ''),
                            bairro=data.get('bairro', ''),
                            localidade=data.get('localidade', ''),
                            uf=data.get('uf', ''),
                            ibge=data.get('ibge', ''),
                            gia=data.get('gia', ''),
                            ddd=data.get('ddd', ''),
                            siafi=data.get('siafi', '')
                        )
                    
                    return None
                    
        except asyncio.TimeoutError:
            print(f"Timeout ao buscar CEP {clean_cep}")
            return None
        except aiohttp.ClientError as e:
            print(f"Erro de conexão ao buscar CEP {clean_cep}: {e}")
            return None
        except Exception as e:
            print(f"Erro inesperado ao buscar CEP {clean_cep}: {e}")
            return None
    
    @staticmethod
    def fetch_address_by_cep_sync(cep: str) -> Optional[AddressData]:
        """
        Busca endereço por CEP na API ViaCEP (síncrono)
        
        Args:
            cep: CEP para busca
            
        Returns:
            AddressData se encontrado, None se erro
        """
        try:
            loop = asyncio.get_event_loop()
        except RuntimeError:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
        
        return loop.run_until_complete(
            ViaCEPService.fetch_address_by_cep(cep)
        )
    
    @staticmethod
    async def validate_and_fetch_address(cep: str) -> tuple[bool, Optional[AddressData], str]:
        """
        Valida CEP e busca endereço com tratamento de erros
        
        Args:
            cep: CEP para validação e busca
            
        Returns:
            Tupla (is_valid, address_data, error_message)
        """
        if not cep:
            return False, None, "CEP é obrigatório"
        
        if not ViaCEPService.validate_cep_format(cep):
            return False, None, "CEP deve conter exatamente 8 dígitos"
        
        address_data = await ViaCEPService.fetch_address_by_cep(cep)
        
        if address_data is None:
            return False, None, "CEP não encontrado ou inválido"
        
        # Verifica se o endereço tem informações mínimas
        if not address_data.logradouro or not address_data.bairro or not address_data.localidade:
            return False, None, "CEP encontrado mas com informações incompletas"
        
        return True, address_data, ""
    
    @staticmethod
    def validate_and_fetch_address_sync(cep: str) -> tuple[bool, Optional[AddressData], str]:
        """
        Versão síncrona da validação e busca de endereço
        
        Args:
            cep: CEP para validação e busca
            
        Returns:
            Tupla (is_valid, address_data, error_message)
        """
        try:
            loop = asyncio.get_event_loop()
        except RuntimeError:
            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
        
        return loop.run_until_complete(
            ViaCEPService.validate_and_fetch_address(cep)
        )
    
    @staticmethod
    def format_address_for_display(address_data: AddressData) -> str:
        """
        Formata endereço para exibição
        
        Args:
            address_data: Dados do endereço
            
        Returns:
            Endereço formatado para exibição
        """
        parts = []
        
        if address_data.logradouro:
            parts.append(address_data.logradouro)
        
        if address_data.bairro:
            parts.append(address_data.bairro)
        
        if address_data.localidade and address_data.uf:
            parts.append(f"{address_data.localidade}/{address_data.uf}")
        
        if address_data.cep:
            formatted_cep = ViaCEPService.format_cep(address_data.cep)
            if formatted_cep:
                parts.append(f"CEP: {formatted_cep}")
        
        return " - ".join(parts)


# Funções de conveniência para uso direto
async def get_address_by_cep(cep: str) -> Optional[AddressData]:
    """
    Função de conveniência para busca assíncrona de endereço
    
    Args:
        cep: CEP para busca
        
    Returns:
        AddressData se encontrado, None caso contrário
    """
    return await ViaCEPService.fetch_address_by_cep(cep)


def get_address_by_cep_sync(cep: str) -> Optional[AddressData]:
    """
    Função de conveniência para busca síncrona de endereço
    
    Args:
        cep: CEP para busca
        
    Returns:
        AddressData se encontrado, None caso contrário
    """
    return ViaCEPService.fetch_address_by_cep_sync(cep)

File: frontend/src/services/viacep-service.js
/**
 * Serviço ViaCEP - Cliente (JavaScript)
 * Integração com API ViaCEP para busca de endereços por CEP
 */

/**
 * URL base da API ViaCEP
 */
const VIACEP_BASE_URL = 'https://viacep.com.br/ws';

/**
 * Timeout padrão para requisições (em milissegundos)
 */
const DEFAULT_TIMEOUT = 10000;

/**
 * Estados brasileiros válidos
 */
export const BRAZILIAN_STATES = new Set([
    'AC', 'AL', 'AP', 'AM', 'BA', 'CE', 'DF', 'ES', 'GO', 
    'MA', 'MT', 'MS', 'MG', 'PA', 'PB', 'PR', 'PE', 'PI', 
    'RJ', 'RN', 'RS', 'RO', 'RR', 'SC', 'SP', 'SE', 'TO'
]);

/**
 * Classe para dados de endereço
 */
export class AddressData {
    constructor(data = {}) {
        this.cep = data.cep || '';
        this.logradouro = data.logradouro || '';
        this.complemento = data.complemento || '';
        this.bairro = data.bairro || '';
        this.localidade = data.localidade || '';
        this.uf = data.uf || '';
        this.ibge = data.ibge || '';
        this.gia = data.gia || '';
        this.ddd = data.ddd || '';
        this.siafi = data.siafi || '';
        this.erro = data.erro || false;
    }

    /**
     * Verifica se o endereço é válido
     * @returns {boolean} True se válido
     */
    isValid() {
        return !this.erro && this.cep && this.logradouro && this.bairro && this.localidade && this.uf;
    }

    /**
     * Formata endereço para exibição
     * @returns {string} Endereço formatado
     */
    getFormattedAddress() {
        if (!this.isValid()) return '';
        
        const parts = [this.logradouro, this.bairro, this.localidade, this.uf];
        return parts.filter(part => part).join(', ');
    }

    /**
     * Converte para objeto simples
     * @returns {object} Dados do endereço
     */
    toObject() {
        return {
            cep: this.cep,
            street: this.logradouro,
            complement: this.complemento,
            neighborhood: this.bairro,
            city: this.localidade,
            state: this.uf,
            ibge: this.ibge,
            ddd: this.ddd
        };
    }
}

/**
 * Remove formatação do CEP
 * @param {string} cep - CEP com ou sem formatação
 * @returns {string} CEP apenas com números
 */
export function cleanCEP(cep) {
    if (!cep) return '';
    return cep.toString().replace(/\D/g, '');
}

/**
 * Formata CEP no padrão XXXXX-XXX
 * @param {string} cep - CEP apenas com números
 * @returns {string} CEP formatado
 */
export function formatCEP(cep) {
    const clean = cleanCEP(cep);
    
    if (clean.length !== 8) return clean;
    
    return `${clean.slice(0, 5)}-${clean.slice(5)}`;
}

/**
 * Valida formato do CEP
 * @param {string} cep - CEP para validação
 * @returns {boolean} True se válido
 */
export function validateCEPFormat(cep) {
    const clean = cleanCEP(cep);
    
    // CEP deve ter exatamente 8 dígitos
    if (clean.length !== 8) return false;
    
    // CEP não pode ser todos zeros
    if (clean === '00000000') return false;
    
    // Verifica se é um número válido
    return /^\d{8}$/.test(clean);
}

/**
 * Aplica máscara de CEP em tempo real
 * @param {string} value - Valor atual do input
 * @returns {string} Valor com máscara aplicada
 */
export function applyCEPMask(value) {
    const clean = cleanCEP(value);
    
    if (clean.length <= 5) return clean;
    
    return `${clean.slice(0, 5)}-${clean.slice(5, 8)}`;
}

/**
 * Classe principal do serviço ViaCEP
 */
export class ViaCEPService {
    constructor(options = {}) {
        this.timeout = options.timeout || DEFAULT_TIMEOUT;
        this.baseUrl = options.baseUrl || VIACEP_BASE_URL;
    }

    /**
     * Busca endereço por CEP
     * @param {string} cep - CEP para busca
     * @returns {Promise<AddressData>} Dados do endereço
     */
    async fetchAddressByCEP(cep) {
        const cleanCep = cleanCEP(cep);
        
        if (!validateCEPFormat(cleanCep)) {
            throw new Error('CEP inválido: deve conter 8 dígitos');
        }

        const url = `${this.baseUrl}/${cleanCep}/json/`;
        
        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);
            
            const response = await fetch(url, {
                method: 'GET',
                headers: {
                    'Accept': 'application/json',
                    'Content-Type': 'application/json'
                },
                signal: controller.signal
            });
            
            clearTimeout(timeoutId);
            
            if (!response.ok) {
                throw new Error(`Erro HTTP: ${response.status} - ${response.statusText}`);
            }
            
            const data = await response.json();
            
            if (data.erro) {
                throw new Error('CEP não encontrado');
            }
            
            return new AddressData(data);
            
        } catch (error) {
            if (error.name === 'AbortError') {
                throw new Error('Timeout: A busca por CEP demorou muito para responder');
            }
            
            if (error.message.includes('Failed to fetch')) {
                throw new Error('Erro de conexão: Verifique sua internet');
            }
            
            throw error;
        }
    }

    /**
     * Busca múltiplos CEPs
     * @param {string[]} ceps - Array de CEPs
     * @returns {Promise<AddressData[]>} Array de dados de endereço
     */
    async fetchMultipleCEPs(ceps) {
        const promises = ceps.map(cep => 
            this.fetchAddressByCEP(cep).catch(error => ({
                error: error.message,
                cep: cleanCEP(cep)
            }))
        );
        
        return Promise.all(promises);
    }

    /**
     * Busca endereço com retry automático
     * @param {string} cep - CEP para busca
     * @param {number} maxRetries - Número máximo de tentativas
     * @returns {Promise<AddressData>} Dados do endereço
     */
    async fetchAddressWithRetry(cep, maxRetries = 3) {
        let lastError;
        
        for (let attempt = 1; attempt <= maxRetries; attempt++) {
            try {
                return await this.fetchAddressByCEP(cep);
            } catch (error) {
                lastError = error;
                
                if (attempt < maxRetries) {
                    // Aguarda antes de tentar novamente (backoff exponencial)
                    await new Promise(resolve => 
                        setTimeout(resolve, Math.pow(2, attempt) * 1000)
                    );
                }
            }
        }
        
        throw lastError;
    }
}

/**
 * Instância padrão do serviço
 */
export const viacepService = new ViaCEPService();

/**
 * Função de conveniência para busca rápida
 * @param {string} cep - CEP para busca
 * @returns {Promise<AddressData>} Dados do endereço
 */
export async function searchCEP(cep) {
    return viacepService.fetchAddressByCEP(cep);
}

/**
 * Hook para busca de CEP em React
 * @param {string} cep - CEP para busca
 * @returns {object} Estado da busca
 */
export function useCEPSearch(cep) {
    const [state, setState] = React.useState({
        loading: false,
        data: null,
        error: null,
        isValid: false
    });

    React.useEffect(() => {
        if (!cep || !validateCEPFormat(cep)) {
            setState({
                loading: false,
                data: null,
                error: null,
                isValid: false
            });
            return;
        }

        setState(prev => ({ ...prev, loading: true, error: null }));

        searchCEP(cep)
            .then(data => {
                setState({
                    loading: false,
                    data,
                    error: null,
                    isValid: data.isValid()
                });
            })
            .catch(error => {
                setState({
                    loading: false,
                    data: null,
                    error: error.message,
                    isValid: false
                });
            });
    }, [cep]);

    return state;
}

/**
 * Validador de CEP para formulários
 * @param {string} cep - CEP para validação
 * @returns {Promise<boolean|string>} True se válido, mensagem de erro se inválido
 */
export async function cepValidator(cep) {
    if (!cep) return 'CEP é obrigatório';
    
    if (!validateCEPFormat(cep)) {
        return 'CEP deve conter 8 dígitos';
    }
    
    try {
        const address = await searchCEP(cep);
        return address.isValid() || 'CEP não encontrado';
    } catch (error) {
        return `Erro ao validar CEP: ${error.message}`;
    }
}

/**
 * Gera mensagem de erro para CEP
 * @param {string} cep - CEP que falhou
 * @param {string} error - Erro ocorrido
 * @returns {string} Mensagem de erro
 */
export function generateCEPErrorMessage(cep, error) {
    const cleanCep = cleanCEP(cep);
    
    if (!cleanCep) {
        return 'CEP é obrigatório';
    }
    
    if (cleanCep.length !== 8) {
        return 'CEP deve conter exatamente 8 dígitos';
    }
    
    if (error.includes('não encontrado')) {
        return 'CEP não encontrado. Verifique se está correto';
    }
    
    if (error.includes('conexão')) {
        return 'Erro de conexão. Tente novamente';
    }
    
    if (error.includes('Timeout')) {
        return 'Busca demorou muito. Tente novamente';
    }
    
    return 'Erro ao buscar CEP. Tente novamente';
}

// Exportação default para uso direto
export default {
    service: viacepService,
    search: searchCEP,
    validate: validateCEPFormat,
    clean: cleanCEP,
    format: formatCEP,
    mask: applyCEPMask,
    validator: cepValidator,
    generateErrorMessage: generateCEPErrorMessage,
    useCEPSearch,
    AddressData,
    ViaCEPService,
    BRAZILIAN_STATES
};